### ISSUE TITLE
Smooth vs Hard Top K Loss

### ISSUE THREAD

**Issue by Ajaz-Ahmad** [mahmoodlab/CLAM#32]
Hello Team,

Thanks for sharing such a great work. I have one query, I have binary problem [positive, negative] of slide classification. I used cross-entropy as instance loss function and I am getting good results. But in your work, I see you are using TopKLoss function.
I want to ask what is the reason behind this switch of loss function [smooth or hard]. Also, in your code I guess you are using softmax activation in the final layer but I am using sigmoid [binary problem]. How can I achieve this switching in my work?
Do I need to replace sigmoid to softmax? 
And lastly, in the loss function script, svm.py line 94 to 98, if I understand correctly
the output from the model is of shape [2,2] where first axis gives probability of tile belonging 0 class and second axis probability of tile belonging to 1 class where 0 means normal and 1 mean abnormal. When smooth we compute loss for abnormal tiles else we compute for normal?
Please let me know if my understanding is correct. Also, do let me know why we switch on axis?

Kind Regards

**Comment by fedshyvana** [comment]
As explained in the paper, the smooth top-1 loss is used because it has been shown to be more robust than cross-entropy loss when there is label noise (in the case of instance loss, where supervision comes from pseudolabels). For bag-level classification we default to cross-entropy loss. 

Softmax and sigmoid activation are used to turn predicted logits into probability scores and don't really affect optimization (logits are passed to loss functions before applying activation), so just use whatever makes sense. You can use sigmoid for binary if you want. 

I am not familiar with the exact implementation of the loss function in svm.py, but the expected input to the loss function should be N x C (batch-size x number of classes).
